### PR TITLE
feat: integrate Conflux eSpace

### PR DESCRIPTION
**Pull request by Pana**
## Description

Add integration to Conflux eSpace

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
  - [x] I have removed all in development `console.log`s
  - [x] I have removed all commented code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

**Comment by hanszhoou** [comment]
@Pana Please add `conflux` for below configurations to resolve type error:

https://github.com/DimensionDev/Maskbook/blob/develop/packages/web3-shared/evm/utils/chainDetailed.ts#L87
https://github.com/DimensionDev/Maskbook/blob/develop/packages/web3-shared/evm/utils/chainDetailed.ts#L103
https://github.com/DimensionDev/Maskbook/blob/develop/packages/mask/src/plugins/Trader/pipes.ts#70
https://github.com/DimensionDev/Maskbook/blob/develop/packages/mask/src/plugins/Trader/pipes.ts#136
https://github.com/DimensionDev/Maskbook/blob/develop/packages/mask/src/plugins/Trader/apis/trader/index.ts#5
https://github.com/DimensionDev/Maskbook/blob/develop/packages/mask/src/plugins/Trader/constants/0x.ts#3
https://github.com/DimensionDev/Maskbook/blob/develop/packages/mask/src/plugins/Trader/constants/dodo.ts#5
https://github.com/DimensionDev/Maskbook/blob/develop/packages/mask/src/plugins/Trader/constants/openocean.ts#5
https://github.com/DimensionDev/Maskbook/blob/develop/packages/mask/src/plugins/Trader/trader/0x/useTrade.ts#23
https://github.com/DimensionDev/Maskbook/blob/develop/packages/public-api/src/web.ts#182

**Comment by hanszhoou** [comment]
@Pana Under my test, metamask does not support it.

What wallet support conflux? You should also integrate that wallet into mask.

**Review by Jack-Works — APPROVED** [review]

**Comment by Pana** [comment]
@zhouhanseng Metamask now support the Conflux eSpace, you can use this RPC https://evm.confluxrpc.com/
This is a introduction of how to use Metamask with Space https://developer.confluxnetwork.org/guides/en/user_metamask_interact_evmspace

**Comment by hanszhoou** [comment]
@Pana Yes it works. Please hide these two applications here when switch to Conflux, since they haven't been supported on Conflux. Thank you.

image

**Comment by guanbinrui** [comment]
> @Pana Yes it works. Please hide these two applications here when switch to Conflux, since they haven't been supported on Conflux. Thank you.
> 
> image

Hi, @DimensionDev/dev-smartcontract should we deploy our contracts on CFX?

**Comment by hanszhoou** [comment]
@Pana Hi, there're some configurations missed, please check the integrity according to https://github.com/DimensionDev/Maskbook/blob/develop/docs/evm-integration.md

**Comment by Pana** [comment]
@zhouhanseng How to hide that two applications ?

**Comment by hanszhoou** [comment]
@Pana Hello, could you send invitation to me for your repo? Or I will open a PR to update the contract info to support the above applications.

**Comment by hanszhoou** [comment]
@Pana How do I swap ETH for CFX for the first time using https://shuttleflow.io/ since no gas fee to claim the CFX?

image

image

**Comment by hanszhoou** [comment]
@Pana What‘s the address of `Conflux_testnet` native token which need to be added to [web3-constants/evm/token.json](https://github.com/DimensionDev/Maskbook/blob/develop/packages/web3-constants/evm/token.json) ? 

I can't find it on the transfer transaction: https://evmtestnet.confluxscan.net/tx/0x48806cdfa8445f229e1a09068934136de81dbe130e243ade9d13dd4d4e1aa624

image

Is that `0x00...0000`?

**Comment by Pana** [comment]
@zhouhanseng The invitation has been send

**Comment by Pana** [comment]
The claim operation should not need gas fee cause it's sponsored

**Comment by Pana** [comment]
CFX is the native token, it doesn't have a address, just like ETH

**Comment by Pana** [comment]
Just mention, The conflux network have two space: Core and eSpace. The eSpace is a complete EVM compatible chain, which we are integrating with Mask. This is eSpace documentation https://developer.confluxnetwork.org/conflux-doc/docs/EVM-Space/intro_of_evm_space

**Comment by Pana** [comment]
We are integrating Multichain bridge to the eSpace, it should be available in two weeks.

**Comment by Pana** [comment]
This is the space bridge between Core and eSpace https://evm.fluentwallet.com/
You can use it cross CFX from Core to eSpace, or reverse

**Comment by hanszhoou** [comment]
@Pana  
> The claim operation should not need gas fee cause it's sponsored

But there's also an 'Insufficient gas fee' tip, and confirm button is disabled. Is that a bug?

image

**Comment by Pana** [comment]
This transaction was partly sponsored, so it shows an Insufficient gas fee, we will increase the sponsor balance, or you can send me a conflux mainnet address, I will send you some CFX for test

**Comment by Pana** [comment]
You can try again, it should working now

**Comment by hanszhoou** [comment]
@Pana It still not work, please send me some CFX. Thank you.

cfx:aap24v2dn0eu70mzk64m82ry28vr15s48a09mwab3x

**Review by Jack-Works — DISMISSED** [review]

**Comment by Pana** [comment]
@zhouhanseng I have sent you 50 CFX, the hash is 0x4336570604152a3cf036c09f9c67f9c53fe6bd216e9b3491b7d64ffc3a464d10
You need cross them to eSpace to deploy contract  https://evm.fluentwallet.com/

**Comment by hanszhoou** [comment]
Received

File: packages/mask/src/plugins/ITO/SNSAdaptor/CompositionDialog.tsx [hunk 1]
```diff
@@ -59,7 +59,7 @@ export function CompositionDialog(props: CompositionDialogProps) {
     const { attachMetadata, dropMetadata } = useCompositionContext()
 
     const { ITO2_CONTRACT_ADDRESS } = useITOConstants()
-
+    console.log({ ITO2_CONTRACT_ADDRESS })
```
Comment (Jack-Works) [inline packages/mask/src/plugins/ITO/SNSAdaptor/CompositionDialog.tsx]: remove this line

**Comment by hanszhoou** [comment]
@DimensionDev/qa

**Comment by wenluomask** [comment]
1. Conflux‘s Assets and history may not support, but currently no any history in any network;
Screen Shot 2022-03-20 at 12 29 15 AM

2. For the gas fee setting page, if there is no different option to display, you can just make them same; Dashboard transfer page's setting button is unresponsive;
Screen Shot 2022-03-20 at 12 31 05 AM
Screen Shot 2022-03-20 at 12 37 11 AM

3. If you are connecting Mask wallet, Popup page didn't show estimate US$ value;
Screen Shot 2022-03-20 at 12 33 10 AM

4. Cannot transfer NFT in Dashboard since it is unclickable to search contract address;
Screen Shot 2022-03-20 at 12 48 02 AM

5. Tx detail cannot be found;
Screen Shot 2022-03-20 at 12 51 14 AM

Please merge develop into this pr, so I can test lucky drop.

**Comment by wenluomask** [comment]
1,3,5 Fixed.
2. The gas setting page in Dashboard's transfer doesn't show the default gas price;
截屏2022-03-21 下午11 33 23


4. Search contract was available, but once I chose contract, there was no way to choose NFT.（Maybe there are no NFT there?）
截屏2022-03-21 下午11 37 51

5. The icon in network bar will disappear sometimes;
截屏2022-03-21 下午11 36 14

6. Share tweet will show Unknown network if the lucky drop is in Conflux network;
截屏2022-03-22 下午4 22 49

**Comment by hanszhoou** [comment]
@wenluomask fixed.

**Comment by wenluomask** [comment]
Checked.

File: packages/web3-constants/evm/token-list.json [hunk 2]
```diff
@@ -17,6 +17,7 @@
         "Celo": ["https://tokens.r2d2.to/latest/42220/tokens.json"],
         "Fantom": ["https://tokens.r2d2.to/latest/250/tokens.json"],
         "Aurora": ["https://tokens.r2d2.to/latest/1313161554/tokens.json"],
-        "Aurora_Testnet": []
+        "Aurora_Testnet": [],
+        "Conflux": ["https://tokens.r2d2.to/1030/1313161554/tokens.json"]
```
Comment (guanbinrui) [inline packages/web3-constants/evm/token-list.json]: Wrong URL.

File: packages/mask/src/plugins/Wallet/SNSAdaptor/GasSettingDialog/Prior1559GasSetting.tsx [hunk 3]
```diff
@@ -22,6 +24,12 @@ import { useNativeTokenPrice } from '../../hooks/useTokenPrice'
 import type { GasSettingProps } from './types'
 import { useGasSettingStyles } from './useGasSettingStyles'
 
+const minGasPriceOfChain: ChainIdOptionalRecord<string> = {
+    [ChainId.BSC]: '0x12a05f200', // 5
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Wallet/SNSAdaptor/GasSettingDialog/Prior1559GasSetting.tsx]:
```suggestion
    [ChainId.BSC]: pow(10, 8).multipleBy(5), // 5 Gwei
```

File: packages/mask/src/plugins/EVM/UI/Web3State/getAssetsFn.ts [hunk 4]
```diff
@@ -11,15 +11,35 @@ import {
...
+// https://forum.conflux.fun/t/how-to-upvote-debank-proposal-for-conflux-espace-integration/13935
+const TokenUnavailableFromDebankList = [ChainId.Conflux]
+
+const TOKEN_PRICE_COINGECKO_URL_BASE = 'https://api.coingecko.com/api/v3'
+
+async function getNativeTokenPrice(tokenId: string) {
+    const requestPath = `${TOKEN_PRICE_COINGECKO_URL_BASE}/simple/price?ids=${tokenId}&vs_currencies=${CurrencyType.USD}`
+    const prices = await fetch(requestPath).then((r) => r.json() as Promise<CryptoPrice>)
+    return prices
+}
```
Comment (guanbinrui) [inline packages/mask/src/plugins/EVM/UI/Web3State/getAssetsFn.ts]: Recommend to use the [TokenPriceAPI](packages/web3-providers/src/token-price/index.ts).

File: packages/mask/src/plugins/EVM/UI/Web3State/getAssetsFn.ts [hunk 5]
```diff
@@ -66,22 +86,28 @@ export const getFungibleAssetsFn =
...
+        )
+
+        let balanceList: string[] = []
+
+        if (BALANCE_CHECKER_ADDRESS && balanceCheckerContract) {
+            try {
+                balanceList = await balanceCheckerContract?.methods
+                    .balances(
+                        [address],
+                        trustedTokens.map((x) => x.address),
```
Comment (guanbinrui) [inline packages/mask/src/plugins/EVM/UI/Web3State/getAssetsFn.ts:97]: `uniqBy` and `isSameAddress`

File: packages/mask/src/plugins/EVM/UI/Web3State/getAssetsFn.ts [hunk 6]
```diff
@@ -101,6 +127,27 @@ export const getFungibleAssetsFn =
 
         const allTokens = [...assetsFromProvider, ...assetFromChain]
 
+        const getBalance = createGetLatestBalance(context)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/EVM/UI/Web3State/getAssetsFn.ts:125]: Is it impossible to use `Utils` here?

File: packages/mask/src/plugins/Wallet/SNSAdaptor/GasSettingDialog/Prior1559GasSetting.tsx (hunk 3 above)
Comment (hanszhoou) [reply packages/mask/src/plugins/Wallet/SNSAdaptor/GasSettingDialog/Prior1559GasSetting.tsx]: pow(10, 9) done

File: packages/mask/src/plugins/EVM/UI/Web3State/getAssetsFn.ts (hunk 4 above)
Comment (hanszhoou) [reply packages/mask/src/plugins/EVM/UI/Web3State/getAssetsFn.ts]: done

File: packages/web3-constants/evm/token-list.json (hunk 2 above)
Comment (hanszhoou) [reply packages/web3-constants/evm/token-list.json]: done

File: packages/mask/src/plugins/EVM/UI/Web3State/getAssetsFn.ts (hunk 5 above)
Comment (hanszhoou) [reply packages/mask/src/plugins/EVM/UI/Web3State/getAssetsFn.ts:97]: done

File: packages/mask/src/plugins/EVM/UI/Web3State/getAssetsFn.ts (hunk 6 above)
Comment (hanszhoou) [reply packages/mask/src/plugins/EVM/UI/Web3State/getAssetsFn.ts:125]: cycle import

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**